File: Local_Test/benchmark/motion_detection/motion_detection.py
import matplotlib
matplotlib.use('Agg')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime, timedelta 
from scipy.signal import savgol_filter, find_peaks, butter, sosfiltfilt 
import os 
matplotlib.use('Agg')
import sys


'''
The theories are based on the IMWUT 2019 paper, with citation
Feng Zhang, Chenshu Wu, Beibei Wang, Hung-Quoc Lai, Yi Han, and K. J. Ray Liu. 2019. WiDetect: Robust Motion Detection with a Statistical Electromagnetic Model. Proc. ACM Interact. Mob. Wearable Ubiquitous Technol. 3, 3, Article 122 (September 2019), 24 pages. https://doi.org/10.1145/3351280
'''



def load_csi_data(file_path: str) -> pd.DataFrame:
    try:
        df = pd.read_csv(file_path)

        return df
    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
        return None
    except Exception as e:
        print(f"Error loading or processing CSV file: {e}")
        return None
    
def retrieve_float(data_list: str) -> str:
    data_list = data_list.split(',')
    data_list[0] = data_list[0].replace('[','')
    data_list[-1] = data_list[-1].replace(']','')
    data_list = [float(i) for i in data_list]
    return np.array(data_list)
    
def csi_data_parsing(csi: pd.DataFrame) -> pd.DataFrame:
    csi_data_length = csi['len'][0]
    csi['real'] = csi['data'].apply(lambda x: retrieve_float(x)[0:csi_data_length-1:2])
    csi['imaginary'] = csi['data'].apply(lambda x: retrieve_float(x)[1:csi_data_length:2])
    csi['amplitude'] = csi['real']**2 + csi['imaginary']**2
    # c.f. equation (1) in paper, we use square of CSI magnitude
    return csi

def get_subcarrier(csi_data: pd.DataFrame, subcarrier: int) -> np.ndarray:
    try:
        extracted_subcarrier_signal = csi_data['amplitude'].apply(lambda x: x[subcarrier - 1])
    except IndexError:
        print(f"Error: Invalid subcarrier index {subcarrier} is out of bounds for the data.")
        return None
    return extracted_subcarrier_signal.to_numpy()

def sample_auto_covariance(data: np.ndarray, lag: int) -> float:
    T = len(data)
    mean = np.mean(data)
    data_1 = data[lag:T]
    data_2 = data[0:T - lag]
    acf = np.sum((data_1 - mean) * (data_2 - mean)) / T  # equation (2) in paper
    return acf

# Note that the CSI amplitude |H(t, f)|^2 is denoted as \mu(t, f) in the paper following equation (1).

def rho_G(data: np.ndarray, lag: int) -> float:
    ACF_G = sample_auto_covariance(data, lag)
    ACF_0 = sample_auto_covariance(data, 0)
    if ACF_0 == 0:
        return None
    return ACF_G / ACF_0  # Mentioned in Section 3.2, assumption (2)


def motion_detection(
        csi_data: pd.DataFrame,
        threshold: float = 0.2,
        window_size: int = 100,
        step_size: int = 50,
) -> np.ndarray:
    csi_data = csi_data_parsing(csi_data)
    lag = 1
    time_horizon = len(csi_data)
    num_of_windows = (time_horizon - window_size) // step_size + 1
    detection_outcome = np.zeros(num_of_windows)
    num_of_subcarriers = csi_data['len'][0] // 2  # assuming len is even, 234 in our test case
    subacarrier_csi = [
        get_subcarrier(csi_data, i) for i in range(1, num_of_subcarriers + 1)
    ]
    for i in range(num_of_windows):
        start = i * step_size
        end = min(i * step_size + window_size, time_horizon)
        sum_of_subcarrier_estimator = 0
        valid_subcarriers = 0
        estimator_i = 0
        for j in range(num_of_subcarriers):
            rho_j = rho_G(subacarrier_csi[j][start:end], lag)
            if rho_j is not None:
                valid_subcarriers += 1
                sum_of_subcarrier_estimator += rho_j
        if valid_subcarriers == 0:
            print(f"Warning: No valid subcarriers found for window {i}.")
            continue
        if valid_subcarriers > 0:
            estimator_i = sum_of_subcarrier_estimator / num_of_subcarriers
        print(estimator_i)
        if estimator_i > threshold:
            detection_outcome[i] = 1
        else:
            detection_outcome[i] = 0
    return detection_outcome
    


if __name__ == '__main__':
    # Load the CSI data
    try:
        path_name = sys.argv[1]
    except IndexError:
        print("Error: Please provide the file name as a command line argument.")
        sys.exit(1)

    file_path = path_name

    csi_data = load_csi_data(file_path)

    if csi_data is not None:
        # Perform motion detection
        detection_outcome = motion_detection(csi_data, threshold=0.4)

        print('Testing file: {}'.format(file_path))
        print("Detection Outcome:")
        print(detection_outcome)
        print(len(detection_outcome))

File: Local_Test/benchmark/motion_detection/motion_detection.c
#include "motion_detection.h"


CSIData load_csi_data(const char *file_path) {
    CSIData csi = {0};
    FILE *file = fopen(file_path, "r");
    if (!file) {
        printf("Error opening file\n");
        return csi;
    }

    printf("Loading CSI data from %s\n", file_path);
    char line[1024];
    int row = 0;

    int col = 0;
    while (fgets(line, sizeof(line), file) && row < MAX_ROWS) {
        // csi.rows[row].len = len;
        char *token = strtok(line, ",");
        col = 0;
        int real_col = 0;
        int imaginary_col = 0;
        while (token != NULL) {
            if (col % 2 == 0) {
                real_col = atof(token);
            } else {
                imaginary_col = atof(token);
                csi.rows[row].amplitude[col / 2] = real_col * real_col + imaginary_col * imaginary_col;
            }
            token = strtok(NULL, ",");
            col++;
        }
        csi.rows[row].len = col / 2;
        row++;
    }
    csi.num_rows = row;
    fclose(file);
    return csi;
}


float sample_auto_covariance(float *data, int data_len, int lag) {
    if (data_len <= lag){
        return 0;
    }
    long double mean = 0;
    for (int i = 0; i < data_len; i++){
        mean += data[i];
    }
    mean /= data_len;
    long double sum = 0;
    for (int i = lag; i < data_len; i++) {
        sum += (data[i] - mean) * (data[i - lag] - mean);
    }
    return sum / data_len;
}

float rho_G(float *data, int data_len, int lag) {
    long double acf_g = sample_auto_covariance(data, data_len, lag);
    long double acf_0 = sample_auto_covariance(data, data_len, 0);
    if (acf_0 == 0){
        return -1;
    }
    return acf_g / acf_0;
}

int* motion_detection(CSIData *csi_data, float threshold, int window_size, int step_size) {
    int time_horizon = csi_data -> num_rows;
    int num_windows = (time_horizon - window_size) / step_size + 1;
    int *detection_outcome = malloc(num_windows * sizeof(int));
    if (!detection_outcome){
        return NULL;
    }

    int num_subcarriers = csi_data -> rows[0].len;
    printf("Number of subcarriers: %d\n", num_subcarriers);
    printf("Time horizon: %d\n", time_horizon);
    printf("Number of windows: %d\n", num_windows);

    for (int i = 0; i < num_windows; i++){
        int start = i * step_size;
        int end = start + window_size;
        if (end > time_horizon) end = time_horizon;

        float sum_rho = 0;
        int valid_subcarriers = 0;

        for (int j = 0; j < num_subcarriers; j++){
            float window_data[window_size];
            for (int k = start; k < end; k++){
                window_data[k - start] = csi_data -> rows[k].amplitude[j];
            }
            
            float rho = rho_G(window_data, end - start, 1);
            if (rho != -1){
                sum_rho += rho;
                valid_subcarriers++;
            }
        }

        if (valid_subcarriers == 0){
            detection_outcome[i] = 0;
            continue;
        }

        float estimator = sum_rho / num_subcarriers;
        detection_outcome[i] = (estimator > threshold) ? 1 : 0;
    }

    return detection_outcome;
}

int main(){
    CSIData csi = load_csi_data("test.csv");
    int window_size = 100;
    int step_size = 25;
    int *result = motion_detection(&csi, 0.4, window_size, step_size);
    int num_windows = (csi.num_rows - window_size) / step_size + 1;
    for (int i = 0; i < num_windows; i++){
        printf("%d ", result[i]);
    }
    free(result);
    return 0;
}